djautotask/sync_rest.py

```diff
@@ -781,6 +790,225 @@ def _assign_field_data(self, instance, object_data):
         return instance
 
 
+class ServiceCallSynchronizer(BatchQueryMixin, Synchronizer):
+    client_class = api.ServiceCallsAPIClient
+    model_class = models.ServiceCallTracker
+    condition_field_name = 'companyID'
+    last_updated_field = 'lastModifiedDateTime'
+
+    related_meta = {
+        'companyID': (models.Account, 'account'),
+        'companyLocationID':
+            (models.AccountPhysicalLocation, 'location'),
+        'status': (models.ServiceCallStatus, 'status'),
+        'creatorResourceID': (models.Resource, 'creator_resource'),
+        'canceledByResourceID': (models.Resource, 'canceled_by_resource')
+    }
+
+    @property
+    def active_ids(self):
+        active_ids = models.Account.objects.filter(
+            active=True
+        ).values_list('id', flat=True).order_by(self.lookup_key)
+
+        return active_ids
+
+    def _assign_field_data(self, instance, object_data):
+        instance.id = object_data['id']
+        instance.description = object_data.get('description')
+        instance.duration = Decimal(str(object_data.get('duration')))
+        instance.complete = object_data.get('isComplete')
+        instance.create_date_time = object_data.get('createDateTime')
+        instance.start_date_time = object_data.get('startDateTime')
+        instance.end_date_time = object_data.get('endDateTime')
+        instance.canceled_date_time = object_data.get('canceledDateTime')
+        instance.last_modified_date_time = \
+            object_data.get('lastModifiedDateTime')
+
+        self._set_datetime_attribute(instance, 'create_date_time')
+        self._set_datetime_attribute(instance, 'start_date_time')
+        self._set_datetime_attribute(instance, 'end_date_time')
+        self._set_datetime_attribute(instance, 'canceled_date_time')
+        self._set_datetime_attribute(instance, 'last_modified_date_time')
+
+        self.set_relations(instance, object_data)
+
+        return instance
+
+    def create(self, **kwargs):
+        """
+        Make a request to Autotask to create a ServiceCall.
+        """
+
+        body = {
+            'companyID': kwargs['account'].id,
+            'companyLocationID': kwargs['location'].id,
+            'startDateTime': kwargs['start_date_time'],
+            'endDateTime': kwargs['end_date_time'],
+            'status': kwargs['status'].id,
+            'description': kwargs['description'],
```

Comment:
We now have a pattern for mapping the model fields to the API field names via the `EDITABLE_FIELDS` dicts on the models. Since synchronizers already have a reference to their `model_class` it wont be difficult to merge that behaviour  so we don't have to duplicate figuring out the field names for creating AND updating.

Reply:
There are differences between them. Most of all, EDITABLE_FIELDS are restricted, it is not in the case of creation. But, as you said, duplications can be improved a little more at the level of model. Will fix that.

Reply:
Ok, that begs the question then. Why is editable fields restricted? Does it need to be? Who are we restricting?

If the answers to those questions are "ourselves", lets change `EDITABLE_FIELDS` to something like `AUTOTASK_FIELDS` and then add everything we intend to edit or create. How does that sound?

Reply:
Yeap, I strongly agree. :+1:

Reply:
@camsweeney or @kti-cameron, can I ask your idea here? We have for now `EDITABLE_FIELDS` and it decides which fields are possible to be edited. Instead, in the case of introducing `AUTOTASK_FIELDS`, we will use this dictionary just for mapping. (i.e. any fields can be asked for edit if this mappings are working) 

Do you have any recommendation or idea here?

Reply:
@kti-cameron To give more context. The main question is: Is there a purpose for `EDITABLE_FIELDS`? Why do we have this safety net, and can we just get rid of it in favour of a dictionary that just maps db column names to the API field names?

Reply:
The description for the issue where `EDITABLE_FIELDS` was added is at https://git.ops.topleft.team/apps/django-topleft/-/issues/1797 .

So Matt suggested adding something like EDITABLE _FIELDS. I'm not sure exactly what Matt's rationale was for keeping a list of fields we can edit vs fields we can't. But I'd guess it has something to do with avoiding inadvertent updates to AT. 

For my part, I don't think it's a bad idea for us to have to be very specific about what TopLeft is allowed to update in AT.

Reply:
Ok, that makes sense to me. So that isn't really a safety net at all. It's just there to be a field mapping to translate db columns to API field names in an easy way. Its not there to be specific about what TopLeft can update (and even if it was, this would we the wrong place for it as we don't want limitations in TopLeft to have ANY effect on `django-connectwise` or `django-autotask`.)

Ok, sounds good. Lets change `EDITABLE_FIELDS` to `AUTOTASK_FIELDS`

Reply:
No, it is a safety net. Fields not it `EDITABLE_FIELDS` won't be added to the object to update in AT. See https://github.com/KerkhoffTechnologies/django-autotask/blob/f9cdbd6a76cca1b84a71175ad95c1d2c801d7436/djautotask/api_rest.py#L337

But I take your point about django-autotask not being the best place to keep these restrictions.

Reply:
Ok, then I say I 100% disagree with the need for this safety net. If we already have the list of changed fields, if that list is wrong due to some error in TopLeft, then it should be handled in TopLeft. 

BUT this still begs the question why is it there, Is there some piece of the puzzle missing here? Are we expecting the data in `changed_fields.items()` to be untrustworthy? I don't see why we can't take `changed_fields` at face value...

Reply:
I can say that the current safety net has a role to force TopLeft app (using djautask) to pass only edit-allowable field values.

But I can take the point that we can trust `changed_fields`, which is developed by us.

Reply:
No, I am saying its absolutely not ok for django-autotask to be a safety net for TopLeft. We should not develop django-autotask to accomodate TopLeft, we should develop TopLeft to accomodate django-autotask.

Also, I am not making the point `changed_fields` is something we can trust, I am saying there might be a reason we can NOT trust `changed_fields` that I am unaware of, and I am currently looking for that reason.

Reply:
Ok, while I can not find a reason we can't trust changed fields, or a precise reason for the safety net, I don't think editable fields is intended to segregate the logic between creating and editing, I think that was just a name picked at the time. So I think we can leave the EDITABLE_FIELDS name in, and even the `if field in api_entity.EDITABLE_FIELDS: ` line in, but @kti-daniel , you can also add in all the fields we need to use to create a model as well :+1:

Reply:
Okay, I will do. Thanks for the discussion, Sam and Cameron.